File: revitpy/live_client.py
"""
Client for the RevitPy Live Server (JSON-RPC 2.0 over WebSocket inside Revit).

Connection details come from the discovery file the server writes
(``~/.revitpy/live.json``, or ``REVITPY_LIVE_DISCOVERY``)::

    import asyncio
    from revitpy.live_client import LiveClient, call_live

    print(call_live("live/status"))             # one-off, synchronous

    async def main() -> None:
        async with LiveClient() as client:      # one connection, many calls
            result = await client.run_file("C:/scripts/report.py")
            print(result["output"])

    asyncio.run(main())

See ``docs/developer/live-server.md`` for the protocol.
"""

from __future__ import annotations

import asyncio
import json
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any

__all__ = [
    "LiveClient",
    "LiveConnectionInfo",
    "LiveServerError",
    "LiveServerNotFoundError",
    "call_live",
]

SUPPORTED_PROTOCOL = 1


class LiveServerNotFoundError(RuntimeError):
    """No running Live Server could be found."""


class LiveServerError(RuntimeError):
    """The Live Server answered with a JSON-RPC error."""

    def __init__(self, code: int, message: str) -> None:
        super().__init__(f"{message} (code {code})")
        self.code = code
        self.message = message


def _discovery_path() -> Path:
    override = os.environ.get("REVITPY_LIVE_DISCOVERY")
    return Path(override) if override else Path.home() / ".revitpy" / "live.json"


@dataclass(frozen=True)
class LiveConnectionInfo:
    url: str
    token: str
    protocol: int = SUPPORTED_PROTOCOL
    revit_version: str | None = None
    pid: int | None = None

    @classmethod
    def from_discovery(cls, path: Path | None = None) -> LiveConnectionInfo:
        """Read the discovery file written by a running Live Server."""
        path = path or _discovery_path()
        hint = (
            "Start it from the RevitPy ribbon in Revit, or set "
            "start_live_server = true in %APPDATA%\\RevitPy\\settings.ini."
        )
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except FileNotFoundError:
            raise LiveServerNotFoundError(
                f"RevitPy Live Server is not running (no {path}). {hint}"
            ) from None
        except (OSError, ValueError) as exc:
            raise LiveServerNotFoundError(
                f"Unreadable Live Server discovery file {path}: {exc}"
            ) from exc
        if not isinstance(data, dict) or not {"url", "token"} <= data.keys():
            raise LiveServerNotFoundError(f"Invalid Live Server discovery file {path}")
        protocol = int(data.get("protocol", SUPPORTED_PROTOCOL))
        if protocol != SUPPORTED_PROTOCOL:
            raise LiveServerError(
                0,
                f"Live Server speaks protocol {protocol}; this client supports "
                f"{SUPPORTED_PROTOCOL}. Update revitpy on one side.",
            )
        return cls(
            url=str(data["url"]),
            token=str(data["token"]),
            protocol=protocol,
            revit_version=data.get("revit_version"),
            pid=data.get("pid"),
        )


def _connect(url: str, headers: dict[str, str], timeout: float) -> Any:
    """Return an async context manager connecting with *headers*."""
    try:
        from websockets.asyncio.client import connect
    except ImportError:  # websockets < 13
        from websockets.client import connect as legacy_connect

        return legacy_connect(
            url, extra_headers=headers, max_size=None, open_timeout=timeout
        )
    return connect(url, additional_headers=headers, max_size=None, open_timeout=timeout)


class LiveClient:
    """One WebSocket connection to the Live Server; use as ``async with``."""

    def __init__(
        self, info: LiveConnectionInfo | None = None, *, timeout: float = 330.0
    ) -> None:
        self._info = info or LiveConnectionInfo.from_discovery()
        self._timeout = timeout
        self._connection: Any = None
        self._websocket: Any = None
        self._next_id = 0

    @property
    def info(self) -> LiveConnectionInfo:
        return self._info

    async def __aenter__(self) -> LiveClient:
        headers = {"Authorization": f"Bearer {self._info.token}"}
        connection = _connect(self._info.url, headers, min(self._timeout, 30.0))
        try:
            self._websocket = await connection.__aenter__()
        except OSError as exc:
            raise LiveServerNotFoundError(
                f"Cannot reach the RevitPy Live Server at {self._info.url} ({exc}). "
                "Is Revit running with the Live Server started?"
            ) from exc
        except Exception as exc:  # handshake rejected (bad/stale token, ...)
            raise LiveServerError(
                0, f"Live Server at {self._info.url} refused the connection: {exc}"
            ) from exc
        self._connection = connection
        return self

    async def __aexit__(self, *exc_info: Any) -> None:
        connection, self._connection, self._websocket = self._connection, None, None
        if connection is not None:
            await connection.__aexit__(*exc_info)

    async def call(self, method: str, params: dict[str, Any] | None = None) -> Any:
        """Send one request and return its ``result``.

        Raises:
            LiveServerError: The server returned a JSON-RPC error.
            TimeoutError: No response within the client timeout.
        """
        if self._websocket is None:
            raise RuntimeError("LiveClient must be used as 'async with LiveClient()'")
        self._next_id += 1
        request_id = self._next_id
        await self._websocket.send(
            json.dumps(
                {
                    "jsonrpc": "2.0",
                    "id": request_id,
                    "method": method,
                    "params": params or {},
                }
            )
        )
        async with asyncio.timeout(self._timeout):
            while True:
                reply = json.loads(await self._websocket.recv())
                if reply.get("id") == request_id:
                    break
        error = reply.get("error")
        if error is not None:
            raise LiveServerError(int(error.get("code", 0)), str(error.get("message")))
        return reply.get("result")

    async def status(self) -> dict[str, Any]:
        return await self.call("live/status")

    async def execute(
        self, code: str, filename: str = "<live>", cwd: str | None = None
    ) -> dict[str, Any]:
        params: dict[str, Any] = {"code": code, "filename": filename}
        if cwd is not None:
            params["cwd"] = cwd
        return await self.call("live/execute", params)

    async def run_file(self, path: str | Path) -> dict[str, Any]:
        return await self.call("live/runFile", {"path": str(Path(path).resolve())})

    async def reload(
        self,
        modules: list[str] | None = None,
        paths: list[str | Path] | None = None,
    ) -> dict[str, Any]:
        return await self.call(
            "live/reload",
            {
                "modules": list(modules or []),
                "paths": [str(Path(p).resolve()) for p in paths or []],
            },
        )

    async def start_debugger(self, port: int = 5678) -> dict[str, Any]:
        return await self.call("debug/start", {"port": port})

    async def list_analyses(self) -> list[str]:
        result = await self.call("bridge/listAnalyses")
        return list(result["analyses"])

    async def analyze(
        self,
        analysis: str,
        elements: list[Any],
        options: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return await self.call(
            "bridge/analyze",
            {"analysis": analysis, "elements": elements, "options": options or {}},
        )


def call_live(
    method: str,
    params: dict[str, Any] | None = None,
    *,
    info: LiveConnectionInfo | None = None,
    timeout: float = 330.0,
) -> Any:
    """Synchronously open a connection, make one call and close it."""

    async def run() -> Any:
        async with LiveClient(info, timeout=timeout) as client:
            return await client.call(method, params)

    return asyncio.run(run())

File: vscode-extension/src/liveClient.ts
/**
 * JSON-RPC 2.0 client for the RevitPy Live Server (WebSocket, bearer token).
 * Mirrors revitpy/live_client.py; see docs/developer/live-server.md.
 */
import { EventEmitter } from 'node:events';
import type { IncomingMessage } from 'node:http';
import WebSocket from 'ws';
import type { RawData } from 'ws';
import type { DebugStartResult, ExecutionResult, ListAnalysesResult, LiveStatus, ReloadResult } from './protocol';

/** The server answered a request with a JSON-RPC error. */
export class JsonRpcError extends Error {
  readonly code: number;
  readonly data?: unknown;
  readonly rpcMessage: string;

  constructor(code: number, message: string, data?: unknown) {
    super(`${message} (code ${code})`);
    this.name = 'JsonRpcError';
    this.code = code;
    this.data = data;
    this.rpcMessage = message;
  }
}

/** The server could not be reached, refused the handshake, or the socket closed. */
export class ConnectionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ConnectionError';
  }
}

/** No response arrived within the request timeout. */
export class RequestTimeoutError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'RequestTimeoutError';
  }
}

export interface ConnectOptions {
  connectTimeoutMs?: number;
  requestTimeoutMs?: number;
}

interface Pending {
  resolve: (value: unknown) => void;
  reject: (error: Error) => void;
  timer: NodeJS.Timeout;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/** One WebSocket connection to the Live Server. Emits `close` (code, reason) once. */
export class LiveClient extends EventEmitter {
  private nextId = 0;
  private readonly pending = new Map<number, Pending>();

  private constructor(
    private readonly socket: WebSocket,
    readonly url: string,
    private readonly requestTimeoutMs: number,
  ) {
    super();
    socket.on('message', (data: RawData) => this.handleMessage(data));
    socket.on('close', (code: number, reason: Buffer) => this.handleClose(code, reason.toString()));
  }

  static connect(url: string, token: string, options: ConnectOptions = {}): Promise<LiveClient> {
    const connectTimeoutMs = options.connectTimeoutMs ?? 30_000;
    const requestTimeoutMs = options.requestTimeoutMs ?? 330_000;

    return new Promise<LiveClient>((resolve, reject) => {
      // Node's `ws` sends no Origin header, which the Live Server requires.
      const socket = new WebSocket(url, {
        headers: { Authorization: `Bearer ${token}` },
        handshakeTimeout: connectTimeoutMs,
        maxPayload: 256 * 1024 * 1024,
      });
      let settled = false;

      const fail = (message: string): void => {
        if (settled) {
          return;
        }
        settled = true;
        socket.terminate();
        reject(new ConnectionError(message));
      };

      // Kept for the socket's lifetime so late errors never go unhandled.
      socket.on('error', (err: Error) => {
        fail(
          `Cannot reach the RevitPy Live Server at ${url} (${err.message}). ` +
            'Is Revit running with the Live Server started?',
        );
      });
      socket.on('unexpected-response', (_req: unknown, res: IncomingMessage) => {
        const status = res.statusCode ?? 0;
        if (status === 401) {
          fail(
            `RevitPy Live Server at ${url} rejected the token (HTTP 401). ` +
              'The discovery file may be stale; restart the Live Server or reconnect.',
          );
        } else {
          fail(`RevitPy Live Server at ${url} refused the connection (HTTP ${status}).`);
        }
      });
      socket.once('open', () => {
        if (settled) {
          return;
        }
        settled = true;
        resolve(new LiveClient(socket, url, requestTimeoutMs));
      });
    });
  }

  get isOpen(): boolean {
    return this.socket.readyState === WebSocket.OPEN;
  }

  call<T = unknown>(method: string, params: Record<string, unknown> = {}, timeoutMs?: number): Promise<T> {
    if (!this.isOpen) {
      return Promise.reject(new ConnectionError('Not connected to the RevitPy Live Server.'));
    }
    const id = ++this.nextId;
    const timeout = timeoutMs ?? this.requestTimeoutMs;

    return new Promise<T>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new RequestTimeoutError(`${method} timed out after ${Math.round(timeout / 1000)} s`));
      }, timeout);
      this.pending.set(id, { resolve: (value) => resolve(value as T), reject, timer });

      this.socket.send(JSON.stringify({ jsonrpc: '2.0', id, method, params }), (err?: Error) => {
        if (err && this.pending.delete(id)) {
          clearTimeout(timer);
          reject(new ConnectionError(`Could not send ${method}: ${err.message}`));
        }
      });
    });
  }

  status(): Promise<LiveStatus> {
    return this.call<LiveStatus>('live/status');
  }

  execute(code: string, filename?: string, cwd?: string): Promise<ExecutionResult> {
    const params: Record<string, unknown> = { code };
    if (filename !== undefined) {
      params.filename = filename;
    }
    if (cwd !== undefined) {
      params.cwd = cwd;
    }
    return this.call<ExecutionResult>('live/execute', params);
  }

  runFile(filePath: string): Promise<ExecutionResult> {
    return this.call<ExecutionResult>('live/runFile', { path: filePath });
  }

  reload(options: { modules?: string[]; paths?: string[] }): Promise<ReloadResult> {
    return this.call<ReloadResult>('live/reload', {
      modules: options.modules ?? [],
      paths: options.paths ?? [],
    });
  }

  startDebugger(port?: number): Promise<DebugStartResult> {
    return this.call<DebugStartResult>('debug/start', port === undefined ? {} : { port });
  }

  async listAnalyses(): Promise<string[]> {
    const result = await this.call<ListAnalysesResult>('bridge/listAnalyses');
    return result.analyses;
  }

  close(): Promise<void> {
    if (this.socket.readyState === WebSocket.CLOSED) {
      return Promise.resolve();
    }
    return new Promise<void>((resolve) => {
      const fallback = setTimeout(() => this.socket.terminate(), 2000);
      this.socket.once('close', () => {
        clearTimeout(fallback);
        resolve();
      });
      this.socket.close(1000, 'client closing');
    });
  }

  private handleMessage(data: RawData): void {
    let message: unknown;
    try {
      message = JSON.parse(data.toString());
    } catch {
      return;
    }
    if (!isRecord(message) || typeof message.id !== 'number') {
      return;
    }
    const pending = this.pending.get(message.id);
    if (!pending) {
      return;
    }
    this.pending.delete(message.id);
    clearTimeout(pending.timer);

    const error = message.error;
    if (error !== undefined && error !== null) {
      const details = isRecord(error) ? error : {};
      const code = typeof details.code === 'number' ? details.code : 0;
      const text = typeof details.message === 'string' ? details.message : 'Unknown error';
      pending.reject(new JsonRpcError(code, text, details.data));
    } else {
      pending.resolve(message.result);
    }
  }

  private handleClose(code: number, reason: string): void {
    const error = new ConnectionError(
      `Connection to the RevitPy Live Server closed (${code}${reason ? `: ${reason}` : ''}).`,
    );
    for (const pending of this.pending.values()) {
      clearTimeout(pending.timer);
      pending.reject(error);
    }
    this.pending.clear();
    this.emit('close', code, reason);
  }
}
